testsuites/CBLTester/CBL_Functional_tests/TestSetup_FunctionalTests/test_replication_filtering.py

```diff
@@ -0,0 +1,585 @@
+import time
+
+import pytest
+import random
+
+from keywords.MobileRestClient import MobileRestClient
+from keywords.utils import add_new_fields_to_doc
+from keywords.utils import log_info
+from CBLClient.Database import Database
+from CBLClient.Replication import Replication
+from libraries.testkit import cluster
+
+
+@pytest.fixture(scope="function")
+def setup_teardown_test(params_from_base_test_setup):
+    cbl_db_name = "cbl_db"
+    base_url = params_from_base_test_setup["base_url"]
+    db = Database(base_url)
+    db_config = db.configure()
+    log_info("Creating db")
+    cbl_db = db.create(cbl_db_name, db_config)
+
+    yield{
+        "db": db,
+        "cbl_db": cbl_db,
+        "cbl_db_name": cbl_db_name
+    }
+
+    log_info("Deleting the db")
+    db.deleteDB(cbl_db)
+
+
+@pytest.mark.sanity
+@pytest.mark.listener
+@pytest.mark.replication
+@pytest.mark.parametrize("num_of_docs", [
+    10,
+    100,
+    1000
+])
+def test_replication_push_filtering(params_from_base_test_setup, num_of_docs):
+    """
+        @summary:
+        1. Create few docs at CBL.
+        2. Replicate using push and pull and verify that all docs are replicated.
+        3. Add new fields to docs and add boolean filter, which allow docs to replicate only if "new_field_1" is set to
+        true.
+        4. Verify SG has new fields added only when "new_field_1" is true
+    """
+    sg_db = "db"
+    sg_admin_url = params_from_base_test_setup["sg_admin_url"]
+    cluster_config = params_from_base_test_setup["cluster_config"]
+    sg_blip_url = params_from_base_test_setup["target_url"]
+    base_url = params_from_base_test_setup["base_url"]
+    sg_config = params_from_base_test_setup["sg_config"]
+    db = params_from_base_test_setup["db"]
+    cbl_db = params_from_base_test_setup["source_db"]
+    sync_gateway_version = params_from_base_test_setup["sync_gateway_version"]
+
+    if sync_gateway_version < "2.5.0":
+        pytest.skip('This test cannnot run with sg version below 2.5')
+    channels = ["ABC"]
+    username = "autotest"
+    password = "password"
+
+    # Reset cluster to ensure no data in system
+    c = cluster.Cluster(config=cluster_config)
+    c.reset(sg_config_path=sg_config)
+
+    # 1. Creating docs in CBL app
+    sg_client = MobileRestClient()
+    sg_client.create_user(url=sg_admin_url, db=sg_db, name=username, password=password, channels=channels)
+    cookie, session = sg_client.create_session(url=sg_admin_url, db=sg_db, name=username)
+    sync_cookie = "{}={}".format(cookie, session)
+    session_header = {"Cookie": sync_cookie}
+    db.create_bulk_docs(num_of_docs, "cbl", db=cbl_db, channels=channels)
+
+    # 2. Replicating docs to SG
+    # Configure replication with push/pull
+    replicator = Replication(base_url)
+    repl_config = replicator.configure(source_db=cbl_db,
```

Comment:
From design, debugging and readability perspective it's better to use it this way, as it replicate and configure does all this thing in one go and in case of failure we have to trace back. The function it is not written with SRP principle.